kgforge/core/archetypes/external_dataset.py

```diff
@@ -0,0 +1,133 @@
+#
+# Blue Brain Nexus Forge is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Blue Brain Nexus Forge is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Blue Brain Nexus Forge. If not, see <https://choosealicense.com/licenses/lgpl-3.0/>.
+
+import json
+import requests
+
+from typing import Dict, Optional, Union, List
+from kgforge.core import Resource
+from kgforge.core.archetypes.read_store import ReadStore
+from kgforge.core.commons.exceptions import QueryingError
+from kgforge.specializations.stores.bluebrain_nexus import BlueBrainNexus
+
+
+class ExternalDataset(ReadStore):
+    """A class to link to external databases, query and search directly on datasets. """
+
+    def __init__(self, model: Optional["Model"] = None,
+                 ) -> None:
+        super().__init__(model)
+
+    def types(self):
+        # TODO: add other datatypes used, for instance, inside the mappings
+        return list(self.model.mappings(self.model.source, False).keys())
+
+    def search(self, resolvers, *filters, **params):
+        """Search within the database.
+
+        :param keep_original: bool
+        """
+        keep_original = params.pop('keep_original', True)
+        unmapped_resources = self._search(resolvers, *filters, **params)
+        if isinstance(self.service, BlueBrainNexus) or keep_original:
+            return unmapped_resources
+        else:
+            # Try to find the type of the resources within the filters
+            resource_type = type_from_filters(*filters)
+            return self.map(unmapped_resources, type_=resource_type)
+
+    @abstractmethod
+    def _search(self):
+        ...
+
+    def sparql(self, query: str, debug: bool = False, limit: Optional[int] = None,
```

Comment:
doesn't have the same signature as the readstore's sparql method. I think you add parameters to the readstore's sparql method, and it should be the same for the external dataset's and the store

Reply:
ah, let me check